worlds/mlss/Client.py

```diff
@@ -0,0 +1,266 @@
+                logger.info("ERROR: The patch file used to create this ROM is not compatible with "
+                            "this client. Double check your client version against the version being "
+                            "used by the generator.")
+                return False
+        except UnicodeDecodeError:
+            return False
+        except bizhawk.RequestFailedError:
+            return False  # Should verify on the next pass
+
+        ctx.game = self.game
+        ctx.items_handling = 0b101
+        ctx.want_slot_data = True
+        ctx.watcher_timeout = 0.125
+        self.rom_slot_name = rom_name
+        name_bytes = ((await bizhawk.read(ctx.bizhawk_ctx, [(0xB0, 16, "ROM")]))[0])
+        player = ((await bizhawk.read(ctx.bizhawk_ctx, [(0xAF, 1, "ROM")]))[0][0])
+        self.player = player + 1
+        name = bytes([byte for byte in name_bytes if byte != 0]).decode("UTF-8")
+        self.player_name = name
+        logger.info(name)
+
+        for i in range(59):
+            self.checked_flags[i] = [9, 10]
+
+        return True
+
+    async def set_auth(self, ctx: BizHawkClientContext) -> None:
+        ctx.auth = self.player_name
+
+    async def game_watcher(self, ctx: BizHawkClientContext) -> None:
+        from CommonClient import logger
+        try:
+            read_state = await bizhawk.read(ctx.bizhawk_ctx, [(0x4564, 59, "EWRAM"),
+                                                              (0x2330, 2, "IWRAM"), (0x3FE0, 1, "IWRAM"), (0x304A, 1, "EWRAM"), (0x304B, 1, "EWRAM"), (0x304C, 4, "EWRAM"), (0x4800, 6, "EWRAM"), (0x4808, 2, "EWRAM"), (0x4407, 1, "EWRAM"), (0x2339, 1, "IWRAM")])
+            flags = read_state[0]
+            room = (read_state[1][1] << 8) + read_state[1][0]
+            shop_init = read_state[2][0]
+            shop_scroll = read_state[3][0] & 0x1F
+            is_buy = (read_state[4][0] != 0)
+            shop_address = (struct.unpack('<I', read_state[5])[0]) & 0xFFFFFF
+            logo = bytes([byte for byte in read_state[6] if byte < 0x70]).decode("UTF-8")
+            received_index = (read_state[7][0] << 8) + read_state[7][1]
+            cackletta = (read_state[8][0] & 0x40)
+            shopping = (read_state[9][0] & 0xF)
+
+            if logo != "MLSSAP":
+                return
+
+            locs_to_send = set()
+            location = 0
+
+            # Checking shop purchases
+            if is_buy:
+                await bizhawk.write(ctx.bizhawk_ctx, [(0x304A, [0x0, 0x0], "EWRAM")])
+                if shop_address != 0x3c0618 and shop_address != 0x3c0684:
+                    location = shop[shop_address][shop_scroll]
+                else:
+                    if shop_init & 0x1 != 0:
+                        location = badge[shop_address][shop_scroll]
+                    else:
+                        location = pants[shop_address][shop_scroll]
+                if location in ctx.server_locations:
+                    locs_to_send.add(location)
```

Comment:
Is this block checking what players buy from the shop in real time? If so, is there a reasonable way to permanently keep a record in the save data of whether something has been purchased? In case someone loses connection without realizing, plays before they remember to connect, etc.

Also, is it safe to write to `0x304A` at all times? `0x304B` may have been nonzero when you previously called `bizhawk.read`, but the player may have closed the shop between then and now.

Reply:
So in this case 0x304A-0x304B are just copies of the shop state written only into ram when you have purchased an item, so they can be overwritten safely since they only need to be read once. Like you said ofc they have to be connected, and if the case is that they arent connected technically this will save only the copy of the last item you bought in the shop, as a small failsafe. In the case of permanent flags for this I would have to creat bitflags for every shop item since they do not exist.

Reply:
I see, so these are reserved spots in memory for "the player purchased this item most recently," and that one item will still get sent if the player connects after buying something while not connected.

That seems reasonable if permanent flags aren't possible or would be a huge pain.